compiler/dyno/lib/resolution/can-pass.cpp

```diff
@@ -977,6 +977,200 @@ CanPassResult CanPassResult::canPass(Context* context,
+    }
+    if (isParam && !other.isParam) {
+      // If a param is required, can't return a non-param.
+      invalidate();
+      return;
+    }
+    // Ensuring that everything can actually be made
+    // into a reference and const will happen later.
+    // leave isRef and isConst as specified.
+    // We could do some checking now, but that might be a bit premature.
+  }
+
+  QualifiedType::Kind toKind() const {
+    if (!isValid) return QualifiedType::UNKNOWN;
+    if (isType) return QualifiedType::TYPE;
+    if (isParam) return QualifiedType::PARAM;
+    if (isConst) {
+      return isRef ? QualifiedType::CONST_REF : QualifiedType::CONST_VAR ;
+    } else {
+      return isRef ? QualifiedType::REF : QualifiedType::VAR ;
+    }
+  }
+
+  bool valid() const { return isValid; }
+};
+
+static QualifiedType findByPassing(Context* context,
+                                   const std::vector<QualifiedType>& types) {
+  for (auto& type : types) {
+    bool fitsOthers = true;
+    for (auto& otherType : types) {
+      if (otherType == type) continue;
+      auto passes = canPass(context, otherType, type);
+      if (!passes.passes() || passes.promotes()) {
+        fitsOthers = false;
+        break;
+      }
+    }
+    if (fitsOthers) return type;
+  }
+  return QualifiedType();
+}
+
+QualifiedType commonType(Context* context,
+                                const std::vector<QualifiedType>& types,
+                                bool useRequiredKind,
+                                QualifiedType::Kind requiredKind) {
+  assert(types.size() > 0);
+
+  // figure out the kind
+  auto properties = KindProperties::fromKind(types.front().kind());
+  for (auto& type : types) {
+    auto kind = type.kind();
+    auto typeProperties = KindProperties::fromKind(kind);
+    properties.combineWith(typeProperties);
+  }
+
+  if (useRequiredKind) {
+    // The caller enforces a particular kind on us. Make sure that the
+    // computed properties line up with the kind.
+    auto requiredProperties = KindProperties::fromKind(requiredKind);
+    requiredProperties.strictCombineWith(properties);
+    properties = requiredProperties;
+  }
+
+  if (!properties.valid()) return QualifiedType();
+  auto bestKind = properties.toKind();
+
+  // Create a new list of types with their kinds adjusted.
+  std::vector<QualifiedType> adjustedTypes;
+  for (const auto& type : types) {
+    const Param* param = type.param();
+    if (bestKind != QualifiedType::PARAM) {
+      // if we've dropped from params to values, clear the params
+      // from all types
+      param = nullptr;
+    }
+    auto adjustedType = QualifiedType(bestKind, type.type(), param);
+    adjustedTypes.push_back(std::move(adjustedType));
+  }
+
+  // Try apply usual coercion rules to find common type
+  auto commonType = findByPassing(context, adjustedTypes);
+  if (!commonType.isUnknown()) return commonType;
+
+  bool paramRequired = useRequiredKind &&
+    requiredKind == QualifiedType::PARAM;
```

Comment:
will this fit on the previous line? Maybe indent this line to line up with `useReqiredKind` ?